File: ddns/provider/aliesa.py
# coding=utf-8
"""
AliESA API
阿里云边缘安全加速(ESA) DNS 解析操作库
@author: NewFuture, GitHub Copilot
"""

from time import strftime
import ipaddress

from ._base import TYPE_JSON, join_domain
from .alidns import AliBaseProvider


class AliesaProvider(AliBaseProvider):
    """阿里云边缘安全加速(ESA) DNS Provider"""

    endpoint = "https://esa.cn-hangzhou.aliyuncs.com"
    api_version = "2024-09-10"  # ESA API版本
    content_type = TYPE_JSON
    remark = "Managed by DDNS %s" % strftime("%Y-%m-%d %H:%M:%S")

    def _query_zone_id(self, domain):
        # type: (str) -> str | None
        """
        查询站点ID
        https://help.aliyun.com/zh/edge-security-acceleration/esa/api-esa-2024-09-10-listsites
        """
        res = self._request(method="GET", action="ListSites", SiteName=domain, PageSize=500)
        sites = res.get("Sites", [])

        for site in sites:
            if site.get("SiteName") == domain:
                site_id = site.get("SiteId")
                self.logger.debug("Found site ID %s for domain %s", site_id, domain)
                return site_id

        self.logger.error("Site not found for domain: %s", domain)
        return None

    def _query_record(self, zone_id, subdomain, main_domain, record_type, line, extra):
        # type: (str, str, str, str, str | None, dict) -> dict | None
        """
        查询DNS记录
        https://help.aliyun.com/zh/edge-security-acceleration/esa/api-esa-2024-09-10-listrecords
        """
        full_domain = join_domain(subdomain, main_domain)
        res = self._request(
            method="GET",
            action="ListRecords",
            SiteId=int(zone_id),
            RecordName=full_domain,
            # AliESA 只有 A/AAAA 记录类型表示 有不确定数量的IPv4 和 不确定数量的IPv6
            Type='A/AAAA',
            RecordMatchType="exact",  # 精确匹配
            PageSize=100,
        )

        records = res.get("Records", [])
        if len(records) == 0:
            self.logger.warning("No records found for [%s] with %s <%s>", zone_id, full_domain, record_type)
            return None

        # 返回第一个匹配的记录
        record = records[0]
        self.logger.debug("Found record: %s", record)
        return record

    def _create_record(self, zone_id, subdomain, main_domain, value, record_type, ttl, line, extra):
        # type: (str, str, str, str, str, int | str | None, str | None, dict) -> bool
        """
        创建DNS记录
        https://help.aliyun.com/zh/edge-security-acceleration/esa/api-esa-2024-09-10-createrecord
        """
        full_domain = join_domain(subdomain, main_domain)
        extra["Comment"] = extra.get("Comment", self.remark)
        extra["BizName"] = extra.get("BizName", "web")
        extra["Proxied"] = extra.get("Proxied", True)
        data = self._request(
            method="POST",
            action="CreateRecord",
            SiteId=int(zone_id),
            RecordName=full_domain,
            Type=self._get_type(record_type),
            Data={"Value": value},
            Ttl=ttl or 1,
            **extra
        )  # fmt: skip

        if data and data.get("RecordId"):
            self.logger.info("Record created: %s", data)
            return True

        self.logger.error("Failed to create record: %s", data)
        return False
    
    def extract_unique_ip_addresses(self,input_string):
        """
        从逗号分隔的字符串中提取第一个有效的 IPv4 地址和第一个有效的 IPv6 地址。

        :param input_string: 包含潜在 IP 地址的字符串
        :return: 一个元组 (first_ipv4, first_ipv6)，如果未找到则对应位置为空字符串
        """
        # 默认为空字符串方便拼接
        first_ipv4 = ""
        first_ipv6 = ""
        
        # 1. 按逗号分割字符串，并去除每个部分前后的空白字符
        candidates = [part.strip() for part in input_string.split(',')]
        
        # 2. 遍历候选列表
        for candidate in candidates:
            # 跳过空字符串
            if not candidate:
                continue
                
            # 3. 尝试将候选者解析为 IP 地址
            try:
                ip_obj = ipaddress.ip_address(candidate)
                
                # 4. 判断 IP 地址类型，并记录第一个出现的地址
                if isinstance(ip_obj, ipaddress.IPv4Address) and first_ipv4 == "":
                    first_ipv4 = str(ip_obj)
                elif isinstance(ip_obj, ipaddress.IPv6Address) and first_ipv6 == "":
                    first_ipv6 = str(ip_obj)
                    
                # 5. 如果两个地址都找到了，可以提前退出循环以提高效率
                if first_ipv4  and first_ipv6 :
                    break
                    
            except ValueError:
                # 如果解析失败，说明不是有效的 IP 地址，继续检查下一个
                continue
        
        # 6. 返回结果元组
        return first_ipv4, first_ipv6
    
    def record_final(self, old_record , new_record):
        # type: (str, str) -> str
        """
        获取最终结果 只保留一个ipv4和一个ipv6
        """
        ipv4old, ipv6old = self.extract_unique_ip_addresses(old_record)
        ipv4new, ipv6new = self.extract_unique_ip_addresses(new_record)
        ipv4final,ipv6final = self.extract_unique_ip_addresses(ipv4new+","+ipv6new+","+ipv4old+","+ipv6old)
        
        final =  ipv4final+","+ipv6final
        # // 去除可能的开头的逗号
        return final.lstrip(',')

    def _update_record(self, zone_id, old_record, value, record_type, ttl, line, extra):
        # type: (str, dict, str, str, int | str | None, str | None, dict) -> bool
        """
        更新DNS记录
        https://help.aliyun.com/zh/edge-security-acceleration/esa/api-esa-2024-09-10-updaterecord
        """
        # EAS只有A/AAAA记录并且可能包含多个ip,如果当前域名只配置ipV4或ipV6类型, 则只更新对应部分,然后另一种记录不变
        # ESA可以配置多个Ip作为回源ip, 所以要提取第一个ipv4地址和ipv6地址(DDNS场景下不可能有多个ip)
        # 获取最终结果 只保留一个ipv4和一个ipv6
        final_value = self.record_final(old_record.get("Data", {}).get("Value"), value)
        
        # 检查是否需要更新
        if (
            old_record.get("Data", {}).get("Value") == final_value
            and old_record.get("RecordType") == self._get_type(record_type)
            and (not ttl or old_record.get("Ttl") == ttl)
        ):
            self.logger.warning("No changes detected, skipping update for record: %s", old_record.get("RecordName"))
            return True

        extra["Comment"] = extra.get("Comment", self.remark)
        extra["Proxied"] = extra.get("Proxied", old_record.get("Proxied"))
        data = self._request(
            method="POST",
            action="UpdateRecord",
            RecordId=old_record.get("RecordId"),
            Data={"Value": final_value},
            Ttl=ttl,
            **extra
        )  # fmt: skip

        if data and data.get("RecordId"):
            self.logger.info("Record updated: %s", data)
            return True

        self.logger.error("Failed to update record: %s", data)
        return False

    def _get_type(self, record_type):
        # type: (str) -> str
        return "A/AAAA" if record_type in ("A", "AAAA") else record_type

File: rust/src/provider/aliesa.rs
use std::collections::BTreeMap;

use serde_json::{Map, Value, json};

use crate::error::{Error, Result};
use crate::http::Method;
use crate::signature::{acs_timestamp, acs3_authorization, request_nonce, sha256_hex};

use super::base::{
    CrudProvider, ProviderContext, RecordRequest, endpoint_host, join_domain, json_parameters,
    numeric_id, value_to_string,
};

pub struct AliesaProvider<'a> {
    context: ProviderContext<'a>,
    zones: BTreeMap<String, String>,
}

impl<'a> AliesaProvider<'a> {
    pub fn new(context: ProviderContext<'a>) -> Result<Self> {
        if context.id.is_empty() || context.token.is_empty() {
            return Err(Error::Config(
                "AliESA access key id and secret must be configured".to_owned(),
            ));
        }
        Ok(Self {
            context,
            zones: BTreeMap::new(),
        })
    }

    fn api(&self, method: Method, action: &str, values: Map<String, Value>) -> Result<Value> {
        let body = if matches!(method, Method::Get) {
            None
        } else {
            Some(serde_json::to_string(&Value::Object(values.clone()))?)
        };
        let query = if matches!(method, Method::Get) {
            values
                .iter()
                .filter_map(|(key, value)| value_to_string(value).map(|value| (key.clone(), value)))
                .collect()
        } else {
            BTreeMap::new()
        };
        let body_hash = sha256_hex(body.as_deref().unwrap_or(""));
        let headers = BTreeMap::from([
            (
                "host".to_owned(),
                endpoint_host(&self.context.endpoint, "AliESA")?,
            ),
            ("content-type".to_owned(), "application/json".to_owned()),
            ("x-acs-action".to_owned(), action.to_owned()),
            ("x-acs-content-sha256".to_owned(), body_hash.clone()),
            ("x-acs-date".to_owned(), acs_timestamp()),
            ("x-acs-signature-nonce".to_owned(), request_nonce()?),
            ("x-acs-version".to_owned(), "2024-09-10".to_owned()),
        ]);
        let authorization = acs3_authorization(
            &self.context.id,
            &self.context.token,
            method.as_str(),
            "/",
            &crate::http::form_encode(&query),
            &headers,
            &body_hash,
        )?;
        let mut headers = headers;
        headers.insert("authorization".to_owned(), authorization);
        let response = self.context.send_json(method, "/", &query, body, headers)?;
        if let Some(code) = response.get("Code").and_then(Value::as_str) {
            return Err(Error::Provider(format!(
                "AliESA API error {code}: {}",
                self.context.logger.mask(
                    response
                        .get("Message")
                        .and_then(Value::as_str)
                        .unwrap_or("unknown error")
                )
            )));
        }
        Ok(response)
    }
}

impl CrudProvider for AliesaProvider<'_> {
    fn context(&self) -> &ProviderContext<'_> {
        &self.context
    }
    fn zone_cache(&mut self) -> &mut BTreeMap<String, String> {
        &mut self.zones
    }
    fn query_zone_id(&mut self, domain: &str) -> Result<Option<String>> {
        let response = self.api(
            Method::Get,
            "ListSites",
            Map::from_iter([
                ("SiteName".to_owned(), json!(domain)),
                ("PageSize".to_owned(), json!(500)),
            ]),
        )?;
        Ok(response
            .get("Sites")
            .and_then(Value::as_array)
            .and_then(|sites| {
                sites
                    .iter()
                    .find(|site| site.get("SiteName").and_then(Value::as_str) == Some(domain))
                    .and_then(|site| site.get("SiteId"))
                    .and_then(value_to_string)
            }))
    }
    fn query_record(
        &mut self,
        zone_id: &str,
        subdomain: &str,
        main_domain: &str,
        request: &RecordRequest<'_>,
    ) -> Result<Option<Value>> {
        let response = self.api(
            Method::Get,
            "ListRecords",
            Map::from_iter([
                (
                    "SiteId".to_owned(),
                    json!(numeric_id(zone_id, "AliESA site id")?),
                ),
                (
                    "RecordName".to_owned(),
                    json!(join_domain(subdomain, main_domain)),
                ),
                (
                    "Type".to_owned(),
                    json!(if matches!(request.record_type, "A" | "AAAA") {
                        "A/AAAA"
                    } else {
                        request.record_type
                    }),
                ),
                ("RecordMatchType".to_owned(), json!("exact")),
                ("PageSize".to_owned(), json!(100)),
            ]),
        )?;
        Ok(response
            .get("Records")
            .and_then(Value::as_array)
            .and_then(|records| records.first().cloned()))
    }
    fn create_record(
        &mut self,
        zone_id: &str,
        subdomain: &str,
        main_domain: &str,
        request: &RecordRequest<'_>,
    ) -> Result<()> {
        let mut values = json_parameters(request);
        values.extend([
            (
                "SiteId".to_owned(),
                json!(numeric_id(zone_id, "AliESA site id")?),
            ),
            (
                "RecordName".to_owned(),
                json!(join_domain(subdomain, main_domain)),
            ),
            (
                "Type".to_owned(),
                json!(if matches!(request.record_type, "A" | "AAAA") {
                    "A/AAAA"
                } else {
                    request.record_type
                }),
            ),
            ("Data".to_owned(), json!({"Value": request.address})),
            ("Ttl".to_owned(), json!(request.ttl.unwrap_or(1))),
        ]);
        values
            .entry("Comment".to_owned())
            .or_insert_with(|| json!("Managed by DDNS"));
        values
            .entry("BizName".to_owned())
            .or_insert_with(|| json!("web"));
        values
            .entry("Proxied".to_owned())
            .or_insert_with(|| json!(true));
        if self
            .api(Method::Post, "CreateRecord", values)?
            .get("RecordId")
            .is_some()
        {
            Ok(())
        } else {
            Err(Error::Provider("AliESA failed to create record".to_owned()))
        }
    }
    fn update_record(
        &mut self,
        _zone_id: &str,
        record: &Value,
        request: &RecordRequest<'_>,
    ) -> Result<()> {
        let unchanged = record.pointer("/Data/Value").and_then(Value::as_str)
            == Some(request.address)
            && request.ttl.is_none_or(|ttl| {
                record.get("Ttl").and_then(Value::as_u64) == Some(u64::from(ttl))
            });
        if unchanged {
            return Ok(());
        }
        let record_id = record
            .get("RecordId")
            .and_then(|value| match value {
                Value::Number(_) => Some(value.clone()),
                Value::String(value) => value.parse::<u64>().ok().map(|value| json!(value)),
                _ => None,
            })
            .ok_or_else(|| Error::Provider("AliESA record has no numeric RecordId".to_owned()))?;
        let mut values = json_parameters(request);
        values.extend([
            ("RecordId".to_owned(), record_id),
            ("Data".to_owned(), json!({"Value": request.address})),
        ]);
        values
            .entry("Comment".to_owned())
            .or_insert_with(|| json!("Managed by DDNS"));
        values
            .entry("Proxied".to_owned())
            .or_insert_with(|| record.get("Proxied").cloned().unwrap_or(Value::Null));
        if let Some(ttl) = request.ttl {
            values.insert("Ttl".to_owned(), json!(ttl));
        }
        if self
            .api(Method::Post, "UpdateRecord", values)?
            .get("RecordId")
            .is_some()
        {
            Ok(())
        } else {
            Err(Error::Provider("AliESA failed to update record".to_owned()))
        }
    }
}
